### ISSUE TITLE
Save pointer to new object in beforeSave

### ISSUE THREAD

**Issue by mtrezza** [parse-community/parse-server#5186]
### Issue Description

It is currently not possible to reference an object in its `beforeSave` hook to create a new object with a pointer to it.

### Steps to reproduce

_Example scenario:_
A user in `User` class should have a user profile in `UserProfile` class associated with it. A user profile has a pointer to the user.

It is currently only possible to create the user profile in the user's `afterSave` hook. But immediately after saving the user, other API calls may look for the user profile which does not exist. It is not a solution to create a user profile if it does not exists, as that can cause race conditions in which multiple user profiles are created. It is also not a solution to set the pointer to the user as `unique` in the mongoDB collection which prevents multiple user profile from being created but also causes data loss, because the information to update the user profile with is lost after the failed call.
 
### Expected Results

It should be possible to create a pointer to an object in its `beforeSave` hook.

### Actual Outcome

Trying to save a user profile causes a timeout in path="/parse/batch".

### Environment Setup

- **Server**
  - parse-server version: 2.8.4
  - Localhost or remote server: Heroku

- **Database**
  - MongoDB version: 3.6.7
  - Localhost or remote server: mLab

### Logs/Trace

-

**Comment by flovilmart** [comment]
This is the expected behavior as the object is not stored in the DB yet and doesn’t have an objectid that is stored. 

Why can’t you use the afterSave for this intent?

**Comment by mtrezza** [comment]
Because after saving the user other methods try to update the user profile. This can happen immediately after sign-up when the user is created.

In some cases the afterSave hook has not created the user profile yet and the user profile cannot be updated.


I tried to solve this by creating a user profile if it doesn’t exist when trying to update it (via cloud code). But that can also cause race conditions when the update cloud code is called multiple times immediately after each other.

**Comment by flovilmart** [comment]
ok; I believe I get the problem, however creating a pointer to an unexisting object is not allowed, and at the time of the save, the object is not saved (by definition). Anything could prevent the object from saving actually.

What do you suggest?

**Comment by acinader** [comment]
@mtrezza, you had the right idea asking this on StackOverflow: https://stackoverflow.com/questions/53321248/how-to-prevent-race-condition-when-creating-a-document-on-parse-server

this is a design issue that you can solve. 

I'm glad to continue helping you on that thread.

**Comment by mtrezza** [comment]
@acinader, thank you, any help appreciated.

My understanding is that there is no solution to the issue. There is always a possible race consition. Hence the issue opened here.

What is your suggestion to solve this?

**Comment by acinader** [comment]
I asked you a question on SO: "so why even bother doing #2? why not count on #4 to do the work."

Danh and I are trying to help you, but I am currently lacking enough information to do that.

**Comment by mtrezza** [comment]
@acinader I already answered this on SO two days ago. Which information specifically do you need?

Again, there seems to be no way to avoid a race condition so I think it is not just a debugging issue.

**Comment by flovilmart** [comment]
While I appreciate all kinds of discussions, this one has nothing to do here @mtrezza =. double posting on multiple forums is never the way to go. Here, I particularly addressed the fact that what you are facing is indeed not a bug. If you wish to open a PR and suggest a 'safe' way to save pointers to unexisting objects feel free to do so. Otherwise, I would appreciate you stick with stack overflow.
Never you mentioned in your description you'd open a question outside. Not being satisfied with the answers on stackoverflow this is not a valid answer.  And yes race condition there is. If you want to avoid races, I would recommend you use atomic updates, or another way to structure your data.. OR use another DB technology like Postgres that offer transactions.

**Comment by mtrezza** [comment]
@flovilmart Agreed. If the community confirms that parse server cannot do this by design, then I will look into opening a PR. I just wanted to make sure that I did not miss a feature of parse server that actually solved this.

**Comment by acinader** [comment]
In this case, since you're not trying to coordinate multiple clients, I think that there are ways to work around the problem with parse server as is.  But, there are definitely times when I could have used a 'blocking after-save hook' where the object is available in its saved state before the client gets a response.  Glad to re-open @mtrezza if you want to work through it.

**Comment by mtrezza** [comment]
OK, I will look into a PR when I have more time.

If anyone interested, the currently used workaround:
* a unique index in mongoDB on the `user` field in `UserProfile` collection
* a Parse.Cloud function that return the user's profile and creates one if non exists yet; to address concurrency, if no user profile was found and creating a user profile fails with Parse error 137 `DuplicateValue`, then the function queries again for an existing user profile.

**Comment by flovilmart** [comment]
@mtrezza what do you consider as a PR? Because I don't want you to start working on a PR if we know ahead of time that we'll reject it.

**Comment by flovilmart** [comment]
Let me challenge that with the following example:

```
Parse.Cloud.beforeSave(Parse.User, async (req) => { 
  const user = req.object;
  const profile = new Parse.Object('UserProfile');
  profile.set({ user });
  await profile.save();

  throw new Error('There was a proble saving the user');
});
```

In the following example, there may be an issue after the profile saves the user's reference. So in this case you'll have a dangling reference on the profile which is not a good idea either.

**Comment by mtrezza** [comment]
I think is would be the current implementation of Parse Server. The PR should address exactly the issue you mention, where only when saving the user succeeds, the user profile should be created with a user reference. The PR would add that logic.

**Comment by flovilmart** [comment]
```
Parse.Cloud.beforeSave(Parse.User, (req) => {
  const user = req.object;
  if (!user.get('profile') && user.isNew()) {
    const profile = new Parse.Object('UserProfile');
    await profile.save();
    user.set({ profile });
  } 
});

Parse.Cloud.afterSave(Parse.User, (req) => {
  const user = req.object;
  if (user.isNew()) {
    user.get('profile').set('user');
    // uncomment return if you want to wait till the profile saved
    /* return */ user.get('profile').save();
  } 
});
```

What is wrong with this approach? In any case, because you associate the profile with the user save, you will need to fetch in from the client. The server is designed to return any mutations that occur in the before/after hooks.

Also, with version > 3 you can wait till the afterSave has completed.

**Comment by mtrezza** [comment]
This approach causes the profile to be without user reference before `afterSave` has completed. So a profile can only be safely accessed through the User class with an include statement.

The conclusion would be to not add a user reference to the profile at all, because it is not guaranteed to exist . But without reference, the `UserProfile` class cannot be queried meaningfully, because it is missing the user reference.

So the requirement is that the profile needs a user reference from the moment it is created.

**Comment by flovilmart** [comment]
well, just trying to give your the proper way of doing it, without breaking anything and with enough guarantees that the profile saves occurs within the user save cycle. If you manage to query a profile in between this cycle, then I’m not sure what to tell you, but again, consider transactions instead. Whatever you are attempting is also wrong as you would be saving the profile to an unexisting user.

**Comment by mtrezza** [comment]
Going back to my suggestion, what speaks against introduction a `profile.safeOnSuccess()` in the `beforeSave` hook?

Behind the curtain, when the `beforeSave` hook saved the user, but before it calls the completion handler, it could save the user profile with the created user as reference.

```
Parse.Cloud.beforeSave(Parse.User, (req) => {
  const user = req.object;
  if (!user.get('profile') && user.isNew()) {
    const profile = new Parse.Object('UserProfile');
    profile.set("user", user);
    profile.saveOnSuccess();
  } 
});
```

The `beforeSave` hook executes behind the curtain:
1. Save user
2. Save profile (with proper user reference and only if saving user succeeded)
3. Call completion handler

**Comment by flovilmart** [comment]
>  what speaks against introducing a profile.safeOnSuccess() in the beforeSave hook?

This is a quite bad solution, If you want to run something *after the object is saved* then use *afterSave*.

Did you even try the solution that I recommend, that work and is reliable? 

After that, what you want / describe, is a transaction. That exactly many operations occur or none. This is an interesting take on transactions, to be able to trigger them from a beforeSave, and add operations to the 'current' transaction, reusing the result of the previous operation (like the objectId saved).

**Comment by flovilmart** [comment]
> The afterSave hook did not complete yet.

Which has been solved for a while as if you return a promise in the afterSave hook, this will wait for the promise to complete before responding to the user.

**Comment by mtrezza** [comment]
But the `signUpInBackground` completion handler in iOS SDK will be called before `afterSave`.

iOS:
```
// Create user
let user = PFUser()
user.username = username
user.password = password

// Sign up user
user.signUpInBackground() { (succeeded, error) in 

    // Update profile
    PFCloud.callFunction(inBackground: "updateProfile", withParameters: [...]) { (response, error) in

        // Fails; cloud function cannot find a profile with the user reference because afterSave hook has not completed.
    }
}
```

Cloud Code:
```
Parse.Cloud.define("updateProfile", function(request, response) {

    // Find user profile
    let query = new Parse.Query("Profile");
    query.equalTo("user", request.user);
    ...
}
```

**Comment by flovilmart** [comment]
No it won’t if you return a promise in your afterSave hook

**Comment by mtrezza** [comment]
@flovilmart That actually solves the issue and makes the PR unnecessary. Took me a while to understand your example, thanks so much for your patience.

**Comment by flovilmart** [comment]
@mtrezza good! However there is still a slight chance that the _User object save fails, and an dangling profile is created, which is not a big deal, as they are easily identifiable as they lack the back reference to the user (but you can always destroy them in batch as they should not be referenced by anyone).
